protocol/bc/validation.go

```diff
@@ -0,0 +1,142 @@
+package bc
+
+import (
+	"chain/errors"
+	"chain/protocol/vm"
+)
+
+type (
+	// ValidChecker can check its validity with respect to a given
+	// validation state.
+	validChecker interface {
+		// CheckValid checks the entry for validity w.r.t. the given
+		// validation state.
+		checkValid(*validationState) error
+	}
+
+	validationState struct {
+		// The ID of the blockchain
+		blockchainID Hash
+
+		// The enclosing transaction object
+		tx *TxEntries
+
+		// The ID of the nearest enclosing entry
+		entryID Hash
+
+		// The source position, for validating ValueSources
+		sourcePos uint64
+
+		// The destination position, for validating ValueDestinations
+		destPos uint64
+	}
+)
+
+var (
+	errBadTimeRange          = errors.New("bad time range")
+	errEmptyResults          = errors.New("transaction has no results")
+	errEntryType             = errors.New("invalid entry type")
+	errMismatchedAssetID     = errors.New("mismatched asset id")
+	errMismatchedBlock       = errors.New("mismatched block")
+	errMismatchedMerkleRoot  = errors.New("mismatched merkle root")
+	errMismatchedPosition    = errors.New("mismatched value source/dest positions")
+	errMismatchedReference   = errors.New("mismatched reference")
+	errMismatchedValue       = errors.New("mismatched value")
+	errMisorderedBlockHeight = errors.New("misordered block height")
+	errMisorderedBlockTime   = errors.New("misordered block time")
+	errNoPrevBlock           = errors.New("no previous block")
+	errNoSource              = errors.New("no source for value")
+	errNonemptyExtHash       = errors.New("non-empty extension hash")
+	errOverflow              = errors.New("arithmetic overflow/underflow")
+	errPosition              = errors.New("invalid source or destination position")
+	errTxVersion             = errors.New("invalid transaction version")
+	errUnbalanced            = errors.New("unbalanced")
+	errUntimelyTransaction   = errors.New("block timestamp outside transaction time range")
+	errVersionRegression     = errors.New("version regression")
+	errWrongBlockchain       = errors.New("wrong blockchain")
+	errZeroTime              = errors.New("timerange has one or two bounds set to zero")
+)
+
+// ValidateTx validates a transaction.
+func ValidateTx(tx *TxEntries, initialBlockID Hash) error {
+	vs := &validationState{
+		blockchainID: initialBlockID,
+	}
+	return tx.checkValid(vs)
+}
+
+// ValidateBlock validates a block and the transactions within.
+// The consensus program is executed only if prev is non-nil and
+// runProg is true.
+func ValidateBlock(b, prev *BlockEntries, initialBlockID Hash, runProg bool) error {
+	if prev == nil {
+		if b.Body.Height != 1 {
+			return errors.WithDetailf(errNoPrevBlock, "height %d", b.Body.Height)
+		}
+	} else {
+		if b.Body.Version < prev.Body.Version {
+			return errors.WithDetailf(errVersionRegression, "previous block verson %d, current block version %d", prev.Body.Version, b.Body.Version)
```

Comment:
Interesting thought. It might be possible to write the validation logic just once, in terms of (admittedly complex) interfaces that can be satisfied by a simple validation-checker on the one hand, and a validation-error-analyzer on the other. (Ask me to show you how I used this approach in my Trustless Digital Escrow demo.)

Agree about pushing this off to a later PR.